sysroot/usr/include/kernel/lib/LinkedList.hpp

```diff
@@ -0,0 +1,249 @@
+/**
+ * @file LinkedList.hpp
+ * @author Keeton Feavel (keetonfeavel@cedarville.edu)
+ * @brief Standard linked list library
+ * @version 0.3
+ * @date 2020-06-17
+ *
+ * @copyright Copyright the Panix Contributors (c) 2021
+ */
+#pragma once
+
+#include <stdint.h>
+#include <stddef.h>
+
+namespace LinkedList {
+
+template <typename T>
+class LinkedListNode {
+public:
+    /**
+     * @brief Construct a new Linked List Node object
+     *
+     */
+    LinkedListNode()
+    : data(0)
+    , next(NULL)
+    , prev(NULL)
+    {
+        // Default constructor
+    }
+    /**
+     * @brief Construct a new Linked List Node object
+     *
+     * @param v Value to be stored
+     */
+    LinkedListNode(T v)
+    : data(v)
+    , next(NULL)
+    , prev(NULL)
+    {
+        // Value constructor
+    }
+    /**
+     * @brief Construct a new Linked List Node object
+     *
+     * @param n Next node in the list
+     * @param p Previous node in the list
+     * @param v Value to be stored
+     */
+    LinkedListNode(T v, LinkedListNode* n, LinkedListNode* p)
+    : data(v)
+    , next(n)
+    , prev(p)
+    {
+        // Complete constructor
+    }
+    /**
+     * @brief Return the data stored by the node
+     *
+     * @return T Stored data
+     */
+    T Data()
+    {
+        return data;
+    }
+    /**
+     * @brief Get the next node in the linked list
+     *
+     * @return LinkedListNode* Pointer to next node
+     */
+    LinkedListNode* Next()
+    {
+        return next;
+    }
+    /**
+     * @brief Get the previous node in the linked list
+     *
+     * @return LinkedListNode* Pointer to the previous node
+     */
+    LinkedListNode* Previous()
+    {
+        return prev;
+    }
+    /**
+     * @brief Set the node's data
+     *
+     */
+    void SetData(T v)
+    {
+        data = v;
+    }
+    /**
+     * @brief Set the node's next pointer
+     *
+     * @param n Pointer to next node
+     */
+    void SetNext(LinkedListNode* n)
+    {
+        next = n;
+    }
+    /**
+     * @brief Set the node's previous pointer
+     *
+     * @param n Pointer to the previous node
+     */
+    void SetPrevious(LinkedListNode* n)
+    {
+        prev = n;
+    }
+
+private:
+    T data;
+    LinkedListNode* next;
+    LinkedListNode* prev;
+    // TODO: Add a size_t index value?
+};
+
+template <typename T>
+class LinkedList {
+public:
+    LinkedList(T val)
+    {
+        head = LinkedListNode<T>(val);
+    }
+    void PushFront(T val)
+    {
+        LinkedListNode<T> next = head;
+        LinkedListNode<T> newHead = new LinkedListNode<T>(val, next, NULL);
```

Comment:
Ah I think this might be a good place to use smart pointers for reference counting. I can work on an implementation if you want.

See [here](https://en.cppreference.com/w/cpp/memory/shared_ptr) for an example of what I'm thinking of.

Reply:
Don’t think we can do that without a standard library.

Reply:
Well I meant I'd work on a version for our own "standard library"

Reply:
That would be nice. Definitely would use that in a lot more places than this, but that's a different issue.